filter.php

```diff
@@ -501,6 +483,50 @@ private function replace($target, $fulltext) : array {
         return [$replacedlink, $replaced];
     }
 
+    /**
+     * Gets the course/page url from the context.
+     *
+     * @param context $context
+     * @return moodle_url $url
+     */
+    private function url_from_context($context) {
+        global $PAGE;
+
+        // Defaults to page url.
+        $url = $PAGE->url;
+
+        // Find course/cm based on context.
+        $course = null;
+        $cm = null;
+
+        if ($context instanceof \context_module) {
+            list($course, $cm) = get_course_and_cm_from_cmid($context->instanceid);
+        }
+
+        if ($context instanceof \context_course) {
+            $course = get_course($context->instanceid);
+        }
+
+        // If loaded via Ajax, guess the URL from the context.
+        if (strpos($url, 'lib/ajax/service.php') !== false) {
+
+            if ($context instanceof \context_module && !empty($course)) {
+                $url = new moodle_url('/course/view.php', ['id' => $course->id]);
```

Comment:
I dont think this is quite right. Here we need to check if we were able to infer the $cm, and use its native URL if found.

Reply:
We should only guess the URL in the context of a context_course

Reply:
Ah you're right I wasn't sure what the URLs were being used for, now I understand they are for the switch optimised/source button.

I've changed up the logic a bit. Sometimes `$cm->get_url()` is null so it uses the course url and then if there is a cm url, it uses that instead.

Ive tested all kinds of scenarios e.g. course, course section, ajax course section, inside module and they are all working locally.